### PR TITLE
Fix literal union type inference

### LINKED ISSUE #10901: Type Inference Regression
**TypeScript Version:**  TypeScript@next (2.1.0-dev.20160912 or 2.1.0-dev.20160913).

Issue likely introduced in 2.1.0-dev.20160912.

**Code**

The example code is a slightly reduced version of an error which appeared as of this morning in DefinitelyTyped/types-2.0 definitions for **d3-array**. See [DT PR 11184](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/11184) which raised the error as part of its complete Travis tests. and [DT PR 11205](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/11205) for a temporary "fix".

```ts

/**
 * Administrivia: JavaScript primitive types and Date
 */
export type Primitive = number | string | boolean | Date;

/**
 * Administrivia: anything with a valueOf(): number method is comparable, so we allow it in numeric operations
 */
interface Numeric {
    valueOf(): number;
}

// Not very useful, but meets Numeric
class NumCoercible {
    public a: number;

    constructor(a: number) {
        this.a = a;
    }
    public valueOf() {
        return this.a;
    }
}

/**
 * Return the min and max simultaneously.
 */
export function extent<T extends Numeric>(array: Array<T | Primitive>): [T | Primitive, T | Primitive] | [undefined, undefined] {
   // do what D3's d3-array module's extent function would do
}


let extentMixed: [Primitive | NumCoercible, Primitive | NumCoercible] | [undefined, undefined];
extentMixed = extent([new NumCoercible(10), 13, '12', true]);
```

**Expected behavior:**

```ts
extentMixed = extent([new NumCoercible(10), 13, '12', true]);
```

Should compile without error inferring the return type correctly to be implicitly given as `[Primitive | NumCoercible, Primitive | NumCoercible] | [undefined, undefined]`

This worked as expected up until likely the 2.1.0-dev.20160912 build.

**Actual behavior:** 

The code _fails, unless_ the type parameter is explicitly set:

```ts
extentMixed = extent<NumCoercible>([new NumCoercible(10), 13, '12', true]);
```
The implicitly given return type of `extent(...)` is otherwise inferred as `[Primitive | Numeric, Primitive | Numeric] | [undefined, undefined]` causing an error.

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #10901.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -7608,11 +7608,18 @@ namespace ts {
                     // itself. When inferring from a type to itself we effectively find all type parameter
                     // occurrences within that type and infer themselves as their type arguments.
                     let matchingTypes: Type[];
-                    for (const t of (<UnionOrIntersectionType>target).types) {
-                        if (typeIdenticalToSomeType(t, (<UnionOrIntersectionType>source).types)) {
+                    for (const t of (<UnionOrIntersectionType>source).types) {
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Need to update the comment on line 7606

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -7608,11 +7608,18 @@ namespace ts {
...
-                        if (typeIdenticalToSomeType(t, (<UnionOrIntersectionType>source).types)) {
+                    for (const t of (<UnionOrIntersectionType>source).types) {
+                        if (typeIdenticalToSomeType(t, (<UnionOrIntersectionType>target).types)) {
                             (matchingTypes || (matchingTypes = [])).push(t);
                             inferFromTypes(t, t);
                         }
+                        else if (t.flags & (TypeFlags.NumberLiteral | TypeFlags.StringLiteral)) {
+                            const b = getBaseTypeOfLiteralType(t);
+                            if (typeIdenticalToSomeType(b, (<UnionOrIntersectionType>target).types)) {
+                                (matchingTypes || (matchingTypes = [])).push(t);
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Push multiple items in one go `.push(t, b);`

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -7608,11 +7608,18 @@ namespace ts {
...
                     // occurrences within that type and infer themselves as their type arguments.
                     let matchingTypes: Type[];
-                    for (const t of (<UnionOrIntersectionType>target).types) {
-                        if (typeIdenticalToSomeType(t, (<UnionOrIntersectionType>source).types)) {
+                    for (const t of (<UnionOrIntersectionType>source).types) {
+                        if (typeIdenticalToSomeType(t, (<UnionOrIntersectionType>target).types)) {
                             (matchingTypes || (matchingTypes = [])).push(t);
                             inferFromTypes(t, t);
                         }
+                        else if (t.flags & (TypeFlags.NumberLiteral | TypeFlags.StringLiteral)) {
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Is the exclusion of `BooleanLiteral` intentional?

Comment (ahejlsberg) [inline src/compiler/checker.ts]: Yes. The `boolean` type is just another name for the union `true | false`, so when `boolean` is included in a union type it is the same as including `true` and `false`. Therefore we need to special handling for the boolean literals. Similar situation for enum types. However, the `number` and `string` types effectively have infinite sets of literal values we can't flatten them and must handle them specially.

File: src/compiler/checker.ts (hunk 2 above)
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Nice.

File: src/compiler/checker.ts (hunk 1 above)
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Will do.

**Comment by RyanCavanaugh** [comment]
:+1:

**Pull request merged**